In [1]:
import numpy as np
import pandas as pd

In [2]:
df=pd.read_csv('freq_new.csv')

In [3]:
df.shape

(17826, 2)

In [4]:
df

Letter   Frequency
0      about  1226734006
1      other   978481319
2      their   782849411
3      first   578161543
4      price   501651226
...      ...         ...
17821  okgle       12711
17822  oigla       12711
17823  ogfal       12711
17824  ofgle       12711
17825  igole       12711

[17826 rows x 2 columns]

In [5]:
df.shape[0]

17826

In [6]:
list(df.iloc[1,0]).count('i')

0

In [7]:
letter='abcdefghijklmnopqrstuvwxyz'
letter_list=list(letter)
len(letter_list)

26

### Ex1.Exercise 1: There is a database of all 5-letter words (e.g. Google Web Trillion Word corpus). Can you identify the frequency of letter usage  from this?A solution to this is available on the internet, for you to check.

In [8]:
for j in range(len(letter_list)):
  df[letter_list[j]]=0
  for i in range(17826):
    df.iloc[i,j+2]=list(df.iloc[i,0]).count(letter_list[j])

In [9]:
df

Letter   Frequency  a  b  c  d  e  f  g  h  ...  q  r  s  t  u  v  w  x  \
0      about  1226734006  1  1  0  0  0  0  0  0  ...  0  0  0  1  1  0  0  0   
1      other   978481319  0  0  0  0  1  0  0  1  ...  0  1  0  1  0  0  0  0   
2      their   782849411  0  0  0  0  1  0  0  1  ...  0  1  0  1  0  0  0  0   
3      first   578161543  0  0  0  0  0  1  0  0  ...  0  1  1  1  0  0  0  0   
4      price   501651226  0  0  1  0  1  0  0  0  ...  0  1  0  0  0  0  0  0   
...      ...         ... .. .. .. .. .. .. .. ..  ... .. .. .. .. .. .. .. ..   
17821  okgle       12711  0  0  0  0  1  0  1  0  ...  0  0  0  0  0  0  0  0   
17822  oigla       12711  1  0  0  0  0  0  1  0  ...  0  0  0  0  0  0  0  0   
17823  ogfal       12711  1  0  0  0  0  1  1  0  ...  0  0  0  0  0  0  0  0   
17824  ofgle       12711  0  0  0  0  1  1  1  0  ...  0  0  0  0  0  0  0  0   
17825  igole       12711  0  0  0  0  1  0  1  0  ...  0  0  0  0  0  0  0  0   

       y  z  
0      0  0  
1      0  0  
2      0  0  
3      0  0  
4      0  0  
...   .. ..  
17821  0  0  
17822  0  0  
17823  0  0  
17824  0  0  
17825  0  0  

[17826 rows x 28 columns]

In [10]:
frequency=[]
for j in letter_list:
  frequency.append(df[j].sum()) 

In [11]:
frequency.index(max(frequency))

0

In [12]:
data=np.array(letter_list+frequency).reshape(2,26).transpose()

In [13]:
len(frequency)

26

In [14]:
df_letter_frequency=pd.DataFrame( data,[i+1 for i in range(26) ], columns=['letter','frequency'])

In [15]:
df_letter_frequency

letter frequency
1       a      8730
2       b      2532
3       c      3934
4       d      3448
5       e      8045
6       f      1570
7       g      2577
8       h      3174
9       i      6334
10      j         0
11      k      2549
12      l      4686
13      m      3472
14      n      5430
15      o      6274
16      p      2739
17      q         0
18      r      6270
19      s      6342
20      t      4799
21      u      3830
22      v         0
23      w         0
24      x         0
25      y      2390
26      z         0

In [17]:
sorted_df_letter_frequency=df_letter_frequency.sort_values('frequency',ascending=False)

In [18]:
sorted_df_letter_frequency

letter frequency
1       a      8730
5       e      8045
19      s      6342
9       i      6334
15      o      6274
18      r      6270
14      n      5430
20      t      4799
12      l      4686
3       c      3934
21      u      3830
13      m      3472
4       d      3448
8       h      3174
16      p      2739
7       g      2577
11      k      2549
2       b      2532
25      y      2390
6       f      1570
23      w         0
24      x         0
17      q         0
22      v         0
10      j         0
26      z         0

In [19]:
for i in range(26):
  print('frequency of ',sorted_df_letter_frequency.iloc[i,0],'is',sorted_df_letter_frequency.iloc[i,1] )

frequency of  a is 8730
frequency of  e is 8045
frequency of  s is 6342
frequency of  i is 6334
frequency of  o is 6274
frequency of  r is 6270
frequency of  n is 5430
frequency of  t is 4799
frequency of  l is 4686
frequency of  c is 3934
frequency of  u is 3830
frequency of  m is 3472
frequency of  d is 3448
frequency of  h is 3174
frequency of  p is 2739
frequency of  g is 2577
frequency of  k is 2549
frequency of  b is 2532
frequency of  y is 2390
frequency of  f is 1570
frequency of  w is 0
frequency of  x is 0
frequency of  q is 0
frequency of  v is 0
frequency of  j is 0
frequency of  z is 0


### Solution:
frequency of  a is 8730 \
frequency of  e is 8045 \
frequency of  s is 6342 \
frequency of  i is 6334 \
frequency of  o is 6274 \
frequency of  r is 6270 \
frequency of  n is 5430 \
frequency of  t is 4799 \
frequency of  l is 4686 \
frequency of  c is 3934 \
frequency of  u is 3830 \
frequency of  m is 3472 \
frequency of  d is 3448 \
frequency of  h is 3174 \
frequency of  p is 2739 \
frequency of  g is 2577 \
frequency of  k is 2549 \
frequency of  b is 2532 \
frequency of  y is 2390 \
frequency of  f is 1570 \
frequency of  w is 0 \
frequency of  x is 0 \
frequency of  q is 0 \
frequency of  v is 0 \
frequency of  j is 0 \
frequency of  z is 0 \

### Ex2.Exercise 2: We would like to use words which are commonly used - let us assume that the wordle target word is picked at random from possibilities on the internet, and is therefore reflected in the word frequency. Assume that a list of such 5 letter words is available, in order of frequency. But we would like to avoid words in which a letter is used more than once. Can you filter out such words from the list given to you? C

In [20]:
for i in range(26):
  df=df[df[letter_list[i]]<2]
df_non_repeating=df  


In [21]:
df_non_repeating

Letter   Frequency  a  b  c  d  e  f  g  h  ...  q  r  s  t  u  v  w  x  \
0      about  1226734006  1  1  0  0  0  0  0  0  ...  0  0  0  1  1  0  0  0   
1      other   978481319  0  0  0  0  1  0  0  1  ...  0  1  0  1  0  0  0  0   
2      their   782849411  0  0  0  0  1  0  0  1  ...  0  1  0  1  0  0  0  0   
3      first   578161543  0  0  0  0  0  1  0  0  ...  0  1  1  1  0  0  0  0   
4      price   501651226  0  0  1  0  1  0  0  0  ...  0  1  0  0  0  0  0  0   
...      ...         ... .. .. .. .. .. .. .. ..  ... .. .. .. .. .. .. .. ..   
17821  okgle       12711  0  0  0  0  1  0  1  0  ...  0  0  0  0  0  0  0  0   
17822  oigla       12711  1  0  0  0  0  0  1  0  ...  0  0  0  0  0  0  0  0   
17823  ogfal       12711  1  0  0  0  0  1  1  0  ...  0  0  0  0  0  0  0  0   
17824  ofgle       12711  0  0  0  0  1  1  1  0  ...  0  0  0  0  0  0  0  0   
17825  igole       12711  0  0  0  0  1  0  1  0  ...  0  0  0  0  0  0  0  0   

       y  z  
0      0  0  
1      0  0  
2      0  0  
3      0  0  
4      0  0  
...   .. ..  
17821  0  0  
17822  0  0  
17823  0  0  
17824  0  0  
17825  0  0  

[17826 rows x 28 columns]

In [22]:
df_non_repeating.shape

(17826, 28)

### Since number of rows in df and df_non_repeating are same that means that in given list of 5 letter words, that is already filtered.

### Now we take some words with repeating letters and without repeating letters and then filter that.

In [23]:
words=['again','algae','apart','arena','aroma','awake','award','basal','canal','carat','fauna','gamma','koala','madam','mamma','nasal','plaza','radar','salad','sauna']

In [24]:
for i in range(20):
  words.append(df_non_repeating.iloc[i,0])

In [25]:
len(words)

40

In [26]:
list_words=pd.DataFrame(words,index=[i+1 for i in range(40) ],columns=['words'])

In [27]:
list_words

words
1   again
2   algae
3   apart
4   arena
5   aroma
6   awake
7   award
8   basal
9   canal
10  carat
11  fauna
12  gamma
13  koala
14  madam
15  mamma
16  nasal
17  plaza
18  radar
19  salad
20  sauna
21  about
22  other
23  their
24  first
25  price
26  email
27  music
28  after
29  links
30  years
31  items
32  group
33  under
34  games
35  could
36  great
37  hotel
38  store
39  terms
40  right

In [28]:
for j in range(26):
  list_words[letter_list[j]]=0
  for i in range(40):
    list_words.iloc[i,j+1]=list(list_words.iloc[i,0]).count(letter_list[j])

In [29]:
list_words

words  a  b  c  d  e  f  g  h  i  ...  q  r  s  t  u  v  w  x  y  z
1   again  2  0  0  0  0  0  1  0  1  ...  0  0  0  0  0  0  0  0  0  0
2   algae  2  0  0  0  1  0  1  0  0  ...  0  0  0  0  0  0  0  0  0  0
3   apart  2  0  0  0  0  0  0  0  0  ...  0  1  0  1  0  0  0  0  0  0
4   arena  2  0  0  0  1  0  0  0  0  ...  0  1  0  0  0  0  0  0  0  0
5   aroma  2  0  0  0  0  0  0  0  0  ...  0  1  0  0  0  0  0  0  0  0
6   awake  2  0  0  0  1  0  0  0  0  ...  0  0  0  0  0  0  1  0  0  0
7   award  2  0  0  1  0  0  0  0  0  ...  0  1  0  0  0  0  1  0  0  0
8   basal  2  1  0  0  0  0  0  0  0  ...  0  0  1  0  0  0  0  0  0  0
9   canal  2  0  1  0  0  0  0  0  0  ...  0  0  0  0  0  0  0  0  0  0
10  carat  2  0  1  0  0  0  0  0  0  ...  0  1  0  1  0  0  0  0  0  0
11  fauna  2  0  0  0  0  1  0  0  0  ...  0  0  0  0  1  0  0  0  0  0
12  gamma  2  0  0  0  0  0  1  0  0  ...  0  0  0  0  0  0  0  0  0  0
13  koala  2  0  0  0  0  0  0  0  0  ...  0  0  0  0  0  0  0  0  0  0
14  madam  2  0  0  1  0  0  0  0  0  ...  0  0  0  0  0  0  0  0  0  0
15  mamma  2  0  0  0  0  0  0  0  0  ...  0  0  0  0  0  0  0  0  0  0
16  nasal  2  0  0  0  0  0  0  0  0  ...  0  0  1  0  0  0  0  0  0  0
17  plaza  2  0  0  0  0  0  0  0  0  ...  0  0  0  0  0  0  0  0  0  1
18  radar  2  0  0  1  0  0  0  0  0  ...  0  2  0  0  0  0  0  0  0  0
19  salad  2  0  0  1  0  0  0  0  0  ...  0  0  1  0  0  0  0  0  0  0
20  sauna  2  0  0  0  0  0  0  0  0  ...  0  0  1  0  1  0  0  0  0  0
21  about  1  1  0  0  0  0  0  0  0  ...  0  0  0  1  1  0  0  0  0  0
22  other  0  0  0  0  1  0  0  1  0  ...  0  1  0  1  0  0  0  0  0  0
23  their  0  0  0  0  1  0  0  1  1  ...  0  1  0  1  0  0  0  0  0  0
24  first  0  0  0  0  0  1  0  0  1  ...  0  1  1  1  0  0  0  0  0  0
25  price  0  0  1  0  1  0  0  0  1  ...  0  1  0  0  0  0  0  0  0  0
26  email  1  0  0  0  1  0  0  0  1  ...  0  0  0  0  0  0  0  0  0  0
27  music  0  0  1  0  0  0  0  0  1  ...  0  0  1  0  1  0  0  0  0  0
28  after  1  0  0  0  1  1  0  0  0  ...  0  1  0  1  0  0  0  0  0  0
29  links  0  0  0  0  0  0  0  0  1  ...  0  0  1  0  0  0  0  0  0  0
30  years  1  0  0  0  1  0  0  0  0  ...  0  1  1  0  0  0  0  0  1  0
31  items  0  0  0  0  1  0  0  0  1  ...  0  0  1  1  0  0  0  0  0  0
32  group  0  0  0  0  0  0  1  0  0  ...  0  1  0  0  1  0  0  0  0  0
33  under  0  0  0  1  1  0  0  0  0  ...  0  1  0  0  1  0  0  0  0  0
34  games  1  0  0  0  1  0  1  0  0  ...  0  0  1  0  0  0  0  0  0  0
35  could  0  0  1  1  0  0  0  0  0  ...  0  0  0  0  1  0  0  0  0  0
36  great  1  0  0  0  1  0  1  0  0  ...  0  1  0  1  0  0  0  0  0  0
37  hotel  0  0  0  0  1  0  0  1  0  ...  0  0  0  1  0  0  0  0  0  0
38  store  0  0  0  0  1  0  0  0  0  ...  0  1  1  1  0  0  0  0  0  0
39  terms  0  0  0  0  1  0  0  0  0  ...  0  1  1  1  0  0  0  0  0  0
40  right  0  0  0  0  0  0  1  1  1  ...  0  1  0  1  0  0  0  0  0  0

[40 rows x 27 columns]

In [30]:
for i in range(26):
  list_words=list_words[list_words[letter_list[i]]<2]

In [31]:
list_words

words  a  b  c  d  e  f  g  h  i  ...  q  r  s  t  u  v  w  x  y  z
21  about  1  1  0  0  0  0  0  0  0  ...  0  0  0  1  1  0  0  0  0  0
22  other  0  0  0  0  1  0  0  1  0  ...  0  1  0  1  0  0  0  0  0  0
23  their  0  0  0  0  1  0  0  1  1  ...  0  1  0  1  0  0  0  0  0  0
24  first  0  0  0  0  0  1  0  0  1  ...  0  1  1  1  0  0  0  0  0  0
25  price  0  0  1  0  1  0  0  0  1  ...  0  1  0  0  0  0  0  0  0  0
26  email  1  0  0  0  1  0  0  0  1  ...  0  0  0  0  0  0  0  0  0  0
27  music  0  0  1  0  0  0  0  0  1  ...  0  0  1  0  1  0  0  0  0  0
28  after  1  0  0  0  1  1  0  0  0  ...  0  1  0  1  0  0  0  0  0  0
29  links  0  0  0  0  0  0  0  0  1  ...  0  0  1  0  0  0  0  0  0  0
30  years  1  0  0  0  1  0  0  0  0  ...  0  1  1  0  0  0  0  0  1  0
31  items  0  0  0  0  1  0  0  0  1  ...  0  0  1  1  0  0  0  0  0  0
32  group  0  0  0  0  0  0  1  0  0  ...  0  1  0  0  1  0  0  0  0  0
33  under  0  0  0  1  1  0  0  0  0  ...  0  1  0  0  1  0  0  0  0  0
34  games  1  0  0  0  1  0  1  0  0  ...  0  0  1  0  0  0  0  0  0  0
35  could  0  0  1  1  0  0  0  0  0  ...  0  0  0  0  1  0  0  0  0  0
36  great  1  0  0  0  1  0  1  0  0  ...  0  1  0  1  0  0  0  0  0  0
37  hotel  0  0  0  0  1  0  0  1  0  ...  0  0  0  1  0  0  0  0  0  0
38  store  0  0  0  0  1  0  0  0  0  ...  0  1  1  1  0  0  0  0  0  0
39  terms  0  0  0  0  1  0  0  0  0  ...  0  1  1  1  0  0  0  0  0  0
40  right  0  0  0  0  0  0  1  1  1  ...  0  1  0  1  0  0  0  0  0  0

[20 rows x 27 columns]

### Hence filtered list of words with non repeating letters.

In [32]:
for i in range(list_words.shape[0]):
  print('',i+1,' ',list_words.iloc[i,0])

 1   about
 2   other
 3   their
 4   first
 5   price
 6   email
 7   music
 8   after
 9   links
 10   years
 11   items
 12   group
 13   under
 14   games
 15   could
 16   great
 17   hotel
 18   store
 19   terms
 20   right


### Hence in the above method we can filter out words with non repeating letters form list of all words  of 5 letters.